### PR TITLE
[MBP-434] Add support for an in-cluster git server

### PR DESCRIPTION
**Pull request by jordigilh**
This PR adds the capability of deploying an in-cluster git server (gitea) when the pattern's spec in `gitconfig.inClusterFork` is set to true.
The added logic will clone the repository defined in `gitConfig.targetRepo` with revision in `gitconfig.targetRevision` into the deployed instance of gitea in-cluster and use that fork URL as the repository source for the application deployed to the gitops operator.  The URL of the forked repository is captured in the `pattern.status.inClusterRepo` field.  

The process also creates a default user stored in a secret in the `gitea` namespace named `gitea-user-credentials` so that users can make changes to the forked repository.

@mbaldessari @beekhof PTAL

### CODE REVIEW

File: controllers/pattern_controller.go [hunk 1]
```diff
@@ -208,19 +190,71 @@ func (r *PatternReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 
 	logOnce("namespace found")
 
+	if qualifiedInstance.Spec.GitConfig.UseInClusterFork {
+		// TargetRepo not defined, instantiate a gitea server and fork origin
```
Comment (beekhof) [inline controllers/pattern_controller.go:194]: Why do we need UseInClusterFork here?
Is that not sensible to imply based on `SourceRepo != nil  && TargetRepo == nil` ?

File: controllers/argo.go [hunk 2]
```diff
@@ -211,19 +211,19 @@ func applicationName(p api.Pattern) string {
 	return fmt.Sprintf("%s-%s", p.Name, p.Spec.ClusterGroupName)
 }
 
-func getApplication(client argoclient.Interface, name string) (error, *argoapi.Application) {
+func getApplication(client argoclient.Interface, name string) (*argoapi.Application, error) {
```
Comment (beekhof) [inline controllers/argo.go:214]: Reason?

File: controllers/pattern_controller.go [hunk 3]
```diff
@@ -511,3 +555,124 @@ func (r *PatternReconciler) actionPerformed(p *api.Pattern, reason string, err e
 	}
 	return r.onReconcileErrorWithRequeue(p, reason, err, nil)
 }
+
+func (r *PatternReconciler) createGitServer() (string, error) {
```
Comment (beekhof) [inline controllers/pattern_controller.go:559]: @mbaldessari I feel like this function, `generateCustomChartValues()`, and many of the `const`s are an argument for an intermediate chart hosted by us that depends on the upstream one (although security folks may still not like that), has a bunch of these defaults, and creates the secrets, routes, etc.  Would be much less work for the operator. WDYT?

File: controllers/pattern_controller.go (hunk 1 above)
Comment (jordigilh) [reply controllers/pattern_controller.go:194]: To make it explicit that you want to have your own in cluster replica. Making it explicit ensures that the user is aware that they will need to fetch the repo's URL from a field in the pattern status and the secret from the gitea namespace.

File: controllers/argo.go (hunk 2 above)
Comment (jordigilh) [reply controllers/argo.go:214]: It's a golang idiom to have the error as the last return variable.

**Comment by adelton** [comment]
Hello, what is the status / plan for this change?

**Comment by mbaldessari** [comment]
Support has been added in operator version 0.0.52

**Pull request closed without merging**